generators/profilegen/src/profile_generator.rb

```diff
@@ -0,0 +1,225 @@
+      check_and_add_operation(obj, operation, resource_provider)
+    end
+  end
+
+  def check_and_add_operation(module_obj, operation, resource_provider)
+    operation_body = module_obj.name + '::' + operation
+    operation_name_ruby = get_model_name(operation)
+
+    if(check_for_availability(@operation_types, operation_name_ruby))
+      raise "#{operation_name} operation is appearing twice for the RP: #{resource_provider}."
+    else
+      @operation_types << { 'operation_name': operation, 'operation_name_ruby': operation_name_ruby, 'operation_body': operation_body}
+    end
+  end
+
+  def generate_model_types(resource_provider, resource_type_version)
+    namespace = @dir_metadata[resource_provider]['namespace'] + + "::#{get_version(resource_type_version)}::Models"
+    namespace = namespace.gsub(/\./, '_' )
+    obj = Module.const_get(namespace)
+    obj.constants.select do |const_name|
+        model_name = const_name.to_s
+        model_body = obj.name + '::' + const_name.to_s
+        method_name = get_model_name(const_name.to_s)
+
+        if(check_for_availability(@model_types, model_name))
+          raise "#{model_name} Model is appearing twice for the RP: #{resource_provider}."
+        else
+          @model_types << {'model_name': model_name, 'model_body': model_body, 'method_name': method_name}
+        end
+    end
+  end
+
+  def get_model_name(model_name)
+    model_name.gsub(/([a-z\d])([A-Z])/,'\1_\2').downcase
+  end
+
+  def get_client_file
+    check_and_create_directory
+    file_name =  "#{@output_dir}/#{@profile_name}/profile_client.rb"
+    File.new(file_name, 'w')
+  end
+
+  def get_module_definition_file
+    check_and_create_directory
+    file_name =  "#{@output_dir}/#{@profile_name}/module_definition.rb"
+    File.new(file_name, 'w')
+  end
+
+  def get_module_file(resource_type_name)
+    check_and_create_directory
+    file_name = get_module_file_name(resource_type_name)
+    @file_names << file_name.sub('.rb', '')
+    complete_file_name = "#{@output_dir}/#{@profile_name}/modules/#{file_name}"
+    File.new(complete_file_name, 'w')
+  end
+
+  def get_spec_file
+    check_and_create_directory
+    file_name =  "#{@output_dir}/#{@profile_name}/azure_mgmt_profile_#{@profile_version}.gemspec"
+    File.new(file_name, 'w')
+  end
+
+  def get_module_file_name(resource_type_name)
+    get_ruby_specific_resource_type_name(resource_type_name).downcase + '_profile_module.rb'
+  end
+
+  def check_and_create_directory
+    Dir.mkdir("#{@output_dir}") unless File.directory?("#{@output_dir}")
+    Dir.mkdir("#{@output_dir}/#{@profile_name}") unless File.directory?("#{@output_dir}/#{@profile_name}")
+    Dir.mkdir("#{@output_dir}/#{@profile_name}/modules") unless File.directory?("#{@output_dir}/#{@profile_name}/modules")
+  end
+
+  def get_ruby_specific_resource_type_name(resource_type_name)
+    resource_type_name.slice(resource_type_name.rindex('.') + 1, resource_type_name.length - 1)
```

Comment:
please document the reason and assumption behind this slice to find out resource types.